feat: add "preset" functionality for npm-scripts config

bryceosterhaus opened the pull request:
Here is my general and basic idea I had for presets. See the README for some more details. https://github.com/liferay/liferay-npm-tools/pull/76/files#diff-025899add8134128ca957d51a3de3f47R96

Essentially a "preset" is a pre-defined `.liferaynpmscriptsrc` config. This also changes the way we merge/override, we now override array config options.

bryceosterhaus commented:
I'm also thinking of removing the whole "eject" functionality. I'm not sure its actually used enough for the upkeep.

diegonvs commented:
Maybe we can try left eject script for debugging

jbalsas commented:
Hey @bryceosterhaus, this sounds good to me! Haven't had time to look into it in detail, but I'd like to give this a try next week once we have everyone back from PTO.

wincent commented:
Code in this looks good. My only concern is that having to create a new NPM package is a pretty heavyweight way of allowing people to override or extend an existing configuration (note: that is not necessarily a bad thing; we might actually *want* a fair bit of friction in order to make it more likely that people will stick to the defaults).

To create an override, you need to do a few "hard" things: pick a meaningful name for it, publish to NPM, update the `Yarn.lock` in the liferay-portal repo etc. Another challenge if we move to TypeScript in this repo is that the dynamic `require` statement using a path computed at runtime is hostile to static analysis.

To analyze what this would mean in a concrete use case, I am currently working on getting the webpack builds that we do in three places in the liferay-portal repo to be managed by liferay-npm-scripts. The use cases are:

1. **frontend-js-web:** we currently explicitly invoke webpack as part of the build, but we'd like to happen automatically.
2. **segments-web:** we are using webpack to work around a circular dependency in a third-party library, currently manually, but we'd like it to be part of the build.
3. **analytics-client-js:** we use webpack to make an independently consumable libray and have no interest in running Babel, the bundler, or doing anything related to soy.

So, "1" is pretty straightforward and wouldn't need any changes to presets or other configuration; we can simply run webpack because we see a "webpack.config.js" file in the repo. "2" as well, I think might be possible to make happen independently of what we do with this PR. "3", however, is an interesting example: in this case we would want to be able to say `liferay-npm-scripts build` and provide configuration that says to skip pretty much everything in the normal build (ie., Babel, soy, bundler) and just do the webpack bit. For this use-case, given this PR, we'd have to create and publish a "liferay-npm-scripts-preset-empty" preset that contains configuration that says to do nothing (separately, would need to decide exactly what that configuration would look like, whether it be a `"babel": false` kind of thing, or `"phases": ["webpack"]`, or any of a number of other possible forms).

I guess, a simple change that would remove a lot of the friction would be to accept a relative path instead of (or in addition to) an NPM package name for the preset config. That would address the concerns with the "hard" things mentioned above.

And that makes me wonder whether even "liferay-npm-scripts-preset-standard" should be a separate NPM package. We could just let it live inside "liferay-npm-scripts", and if you wanted to extend it you would do:

```javascript
// my-fancy-liferay-npm-scripts-preset.js
const standard = require('liferay-npm-scripts/presets/standard');

module.exports = {
  build: {
    ...standard.build,
    inputs: [...standard.build.inputs, 'some-other-input'],
  }
};
```

And you'd use this in you `.liferaynpmscriptsrc` like this:

```
{
  "preset": "./relative/path/to/my-fancy-liferay-npm-scripts-preset.js"
}
```

If we refrained from making liferay-npm-scripts-preset-standard a separate package, we'd avoid management overhead. Among other things, this would enable us to update just "liferay-npm-scripts" instead of having to bump the version of "liferay-npm-scripts" *and* "liferay-npm-scripts-preset-standard" every time we want to make a change to the latter.

bryceosterhaus commented:
When initially discussing this functionality, I know we wanted to in some sense create friction for people overriding the default config. Although, as a user of the tool I think I would enjoy using a local path rather than publishing a new preset every time. So I am really up for either direction, we either purposely make it difficult to override, or we just make it plain in the docs that we don't encourage overriding and allow local presets.

> And that makes me wonder whether even "liferay-npm-scripts-preset-standard" should be a separate NPM package. We could just let it live inside "liferay-npm-scripts",

Overall, I would side with you and go for local presets available and especially because of this and how it makes it easier for publishing our own tool.

@jbalsas I know you liked the idea of adding friction, what do you think about allowing local override?

jbalsas commented:
Hey @wincent, @bryceosterhaus, there's other ways we can add friction. One example is by SF rules, so we don't necessarily need to push friction here, we can do so on the DXP build side.

bryceosterhaus commented:
@jbalsas @wincent updated.

I added the following:
- Added ability to use preset from either local file path or from npm package.
- Renamed the config file from `.liferaynpmscriptsrc` to `liferaynpmscripts.js` (Had to make it a JS file for importing preset)
- Added validation for config to keep from unnamed errors.
- Removed `eject` functionality. (went ahead and removed this to keep our maintenance overhead down since this feature isn't really used at all. I'm up for different thoughts though.)

Let me know what you guys think!

bryceosterhaus commented on packages/liferay-npm-scripts/README.md:
```diff
@@ -79,19 +78,42 @@ liferay-npm-scripts theme build
+## Config
 
-**Note: this is a one-way operation. Once you eject, you can't go back.**
+> Note: as of v2.x the config file was renamed from `.liferaynpmscriptsrc` to `liferaynpmscripts.js`
```
I would be concerned with that being too generic and might cause people to think that is a file geared specifically to liferay's build process, when really its only specific to the npm tools.

What about one of these?
`npmscripts`
`npm-scripts`
`liferayscripts`
`liferay-scripts`

and then a file ending of either `.config.js` or `.conf.js`

I think `npm-scripts.config.js` is pretty clean

wincent replied:
The hyphens definitely help readability (`.liferaynpmscriptsrc` was terrible in that respect). To be honest, any of the suggested names are an improvement. I'd add one more to the list (`liferay-npm-scripts.config.js`) for consideration too.

jbalsas replied:
I like the new options better, but I think I still want to try to win you over to `build.js`.

This is an aggregation of different options found in a `build.gradle` file:

```groovy
buildService {
	apiDir = "../adaptive-media-image-api/src/main/java"
	testDir = "../adaptive-media-image-test/src/testIntegration/java"
}

dependencies {
	compileOnly group: "com.liferay.portal", name: "com.liferay.portal.kernel", version: "default"
	compileOnly project(":apps:static:portal-configuration:portal-configuration-metatype-api")

	testCompile group: "com.liferay.portal", name: "com.liferay.portal.impl", version: "default"
	testCompile project(":core:petra:petra-lang")

	testIntegrationCompile group: "com.liferay.portal", name: "com.liferay.portal.kernel", version: "default"
	testIntegrationCompile project(":apps:adaptive-media:adaptive-media-api")
}

deploy {
	enabled = false
}

liferayOSGi {
	expandCompileInclude = true
}
```

If you look into it, it basically is declaring dependencies and overriding gradle task defaults. It seems awfully close to what we're getting to now, only that we will be doing this for "npm scripts (tasks)".

This is indeed tied to liferay build process and that's the initial philosophy. We're actually already forcing everyone to use `liferay-npm-scripts build` for their frontend build mechanisms (exclusions aside). We also currently only have 4-5 places where we need to override defaults (and that number might shrink), so it shouldn't be that bad.

What do you think?

bryceosterhaus replied:
Personally I just think `build.js` is too generic and doesn't give any idea that there is a dependency required. To push back on the build.gradle, `gradle` is fairly intrinsic to Liferay's build system and require to do *any*(as far as I know...) deployment. Whereas, `liferay-npm-scripts` is an additional tool we can add to our module. If `liferay-npm-scripts` worked without having to declare it as a dependency in our package.json, then I could see the argument for `build.js` but since `liferay-npm-scripts` is an *extra*(required to be listed as dependency) then I think its necessary to make the config explicit in its naming.

Additionally, the extension `.gradle` really helps with the naming of the file. If it was `build.java` I think it would be much more confusing. Since `.js` is so generic and used for a variety of things, I really think it helps keeping the naming explicit.

In the end though, you're the boss, so I am willing to do whatever naming convention you want. 🤷‍♂

jbalsas replied:
> If liferay-npm-scripts worked without having to declare it as a dependency in our package.json, then I could see the argument for build.js but since liferay-npm-scripts is an extra(required to be listed as dependency) then I think its necessary to make the config explicit in its naming.

Technically, it is only listed as a dependency on the root package (which is not really a package), but in the places where it is used it's not, it is a given of the system, just like Gradle

> Additionally, the extension .gradle really helps with the naming of the file. If it was build.java I think it would be much more confusing. Since .js is so generic and used for a variety of things, I really think it helps keeping the naming explicit.

This is my biggest concern too...

> In the end though, you're the boss, so I am willing to do whatever naming convention you want. 🤷‍♂

I really really hope we can do better than that :)

What would you think about `build.npmscripts` or `build.lfrnpmscripts` or something along those lines?

bryceosterhaus replied:
I'd be down for `build.npmscripts`, although it can cause issue for linters and language identification in editors since it doesnt have the .js extension. So I think that would mean these files would sneak past the SF.

> I really really hope we can do better than that :)

hey, bikeshedding needs a kill switch at some point in the end 🚴

jbalsas replied:
> I'd be down for build.npmscripts, although it can cause issue for linters and language identification in editors since it doesnt have the .js extension. So I think that would mean these files would sneak past the SF.

So... `build.npmscripts.js`?

> hey, bikeshedding needs a kill switch at some point in the end 🚴

That's true too, but then we should at least put it to a vote 😂

bryceosterhaus replied:
`build.npmscripts.js` sounds good to me, although I did just remember that inside the config we have keys of `lint` `format` and `build`. Do you think that'll be confusing at all to put lint and format configs inside of a file named build?

jbalsas replied:
Maybe a little, but shouldn't be too much... if you draw the connection, you're just providing  configuration to different tasks.

We currently only have... 4 or 5 instances of this and all of them use it to configure build dependencies. We might not even allow getting out of linting or formatting much, so that might never be a real thing.

I'd be fine with this, but we can wait for @wincent's opinion on this for a final decision. 

If `build.npmscripts.js` is confusing, then the other only options would be the ones you proposed at the beginning. I like this one better for the simmetry it brings, which is always appreciated imho.

wincent replied:
I don't feel super strongly about this so I wasn't going to add any more ⛽️ to the 🔥, but seeing as you asked...

You both make good points. I agree about needing the ".js" extension, and I agree that "build" is a bit misleading for a tool that does things beyond just building. That's why I liked the explicitness of "liferay-npm-scripts.config.js" because it very obviously shows that it is "configuration for liferay-npm-scripts". I don't think that "build.npmscripts.js" is quite as clear, but it is definitely an improvement over "build.js". Another concern with `build.js` is it sounds like a script that I should be able to run (eg. `node build.js`) whereas the variants which end with ".config.js" are much more clearly not intended to be run directly.

So my order of preference would be:

1. liferay-npm-scripts.config.js
2. npm-scripts.config.js
3. npmscripts.config.js
4. build.npmscripts.js
5. build.js
6. .liferaynpmscriptsrc

jbalsas replied:
Ok, taking from @wincent's list, I would remove the ones with `-` because it's a complicated topic in the portal repo, so, let's go with `npmscripts.config.js` and be done with this :)

wincent commented on packages/liferay-npm-scripts/src/utils/validateConfig.js:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Util for validating json for certain keys. If content for each key does not exist,
+ * the process will exit and emit an error with a list of keys missing. Note, this
+ * ultity does not work for keys deeper than the first level.
```
s/ultity/utility/

May be overkill for this use case, but [JSON Schema](https://json-schema.org/) is really nice for defining what constitutes valid JSON data. I wouldn't actually add it at this point because I am allergic to dependencies, but for anything more complex than what we're dealing with here I've found it to be great for getting checks and guard clauses out of the code and into a single validation step.

bryceosterhaus replied:
> I am allergic to dependencies
😂

wincent commented on packages/liferay-npm-scripts/src/utils/validateConfig.js:
```diff
@@ -0,0 +1,33 @@
+			)}.`
+		);
+
+		process.exit(1);
```
One thing to bear in mind, `process.exit()` makes it harder to write tests involving anything that uses it (because you have to stub it to prevent it from crashing the test runner), so generally you're better off using `throw` and keeping one single `process.exit()` call right at the "top" of your entry point.

bryceosterhaus commented:
@wincent @jbalsas updated config file name and also opted to use `throw` instead of `process.exit()`.

wincent commented:
Just rebased this for you @bryceosterhaus seeing as my changes to the yarn.lock in #80 or #81 created a conflict here.

wincent commented:
FYI I'm updating portal to the new version of liferay-npm-scripts in service of https://issues.liferay.com/browse/LPS-95657 — I'll make a point of migrating the few `.liferaynpmscriptsrc` files that we have in the repo over to the new config format:

- modules/apps/portal-search/portal-search-admin-web/.liferaynpmscriptsrc
- modules/apps/asset/asset-list-web/.liferaynpmscriptsrc
- modules/apps/asset/asset-publisher-web/.liferaynpmscriptsrc
- modules/apps/document-library/document-library-web/.liferaynpmscriptsrc
- modules/apps/site-navigation/site-navigation-item-selector-web/.liferaynpmscriptsrc
